Eliminate iteration release

lgu1234 opened the pull request:
Pull request overview
---------------------
Significant computing time is consumed by EnergyPlus for HVAC system simulation. This pull request adds a feature that allows faster calculations of the controls needed to determine the operating point of selected coil and parent configurations.  Previously, in order to calculate part load ratios (PLR) for single speed coils and speed ratios (SR) for multispeed coils, the Regula Falsi method is used to get a control solution by iterating with the entire coil simulation model is called in the residual function. In general, it takes three to five iterations to reach a control solution for each system call. This pull request adds an option for a direct solution of the control problem instead of the iteration approach using Regula Falsi in order to speed up HVAC system simulations.  The full coil model is still used in the end after the PLR or SpeedRatio has been determined by the faster direct method. 

The feature is accessed by using the new PerformancePrecisionTradeoffs input object with the input field called Use Coil Direct Solutions set to Yes.  



### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [x] CI status: all green or justified
 - [ ] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [x] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

lgu1234 committed:
Initial upload

lgu1234 committed:
Merge branch 'develop' into EliminateIteration-Release

lgu1234 commented:
This is an initial check-in for review and comments.

shorowit commented:
@lgu1234 If this is enabled in a given file, will the EMS calling point "InsideHVACSystemIterationLoop" still work? Or would a different EMS calling point need to be used?

lgu1234 commented:
@shorowit There is no EMS code in the code section I modified. It should not impact. HVAC system iteration is a much higher level than I do. My change is restricted at component level.

shorowit commented:
Got it. Thanks for the explanation.

lgu1234 committed:
Remove unused variables

lgu1234 committed:
Merge branch 'develop' into EliminateIteration-Release

lgu1234 committed:
Revise the code and update IO Ref

lgu1234 committed:
Add variable speed DX coils

lgu1234 committed:
Merge branch 'develop' into EliminateIteration-Release

lgu1234 committed:
Solve conflicts

lgu1234 committed:
Merge branch 'develop' into EliminateIteration-Release

rraustad commented on src/EnergyPlus/HVACDXHeatPumpSystem.cc:
```diff
@@ -726,52 +728,58 @@ namespace HVACDXHeatPumpSystem {
+                                    Par(2) = DesOutTemp;
+                                    Par(3) = 1.0; // OnOffAirFlowFrac assume = 1.0 for continuous fan dx system
+                                    Par(5) = double(FanOpMode);
+                                    SolveRoot(Acc, MaxIte, SolFla, PartLoadFrac, DXHeatingCoilResidual, 0.0, 1.0, Par);
```
This is just a thought... so don't put too much weight on this comment. Could this be done in RootSolver so that no code would have to change?

lgu1234 replied:
@rraustad Thanks for a good suggestion. Right now, I am testing all possible coils to see if the direct solution approach can be applied or not. Half work has been through. When I complete all possible coils, I will think how to consolidate the changes. Some changes are beyond my original thought. For example, a linear relationship between system output and part load ratio was proposed. The recent change to handle variable speed coils expands the application with a linear relationship between design output temperature and part load ratio. Therefore, final outcome will be based on all changes.

EnergyArchmage commented on doc/input-output-reference/src/overview/group-simulation-parameters.tex:
```diff
@@ -798,6 +798,94 @@ \subsection{PerformancePrecisionTradeoffs}\label{performanceprecisiontradeoffs}
 
+Although this global parameter is assumed to impact all coils, the real application is applied to a limited number of systems and coils only so far, listed in the follwoing table. Since a linear relationship between system output and part load ratio or speed ratio is expected, an analytical solution is applied to all coils listed in the table. More coils will be allowed when time allows.
+
+The Part Load Ratio (PLR) for a single speed coil or a multiple speed coil at speed 1 is calculated using the equation below
```
I guess this okay for now, but this (equations below here) sort of documentation is better to put into the Engineering Reference.

EnergyArchmage committed:
edit I/O ref for direct coil control method

clean up description in I/O ref for 10X feature for coil model speed up.

minor code changes to use full DataGlobals namespace

EnergyArchmage commented:
I have reviewed this PR and I think it is okay to go in.  Ran with and without new feature in some test files and it seems to work as expected (though during my un-scientific speed testing it was not always any faster...)  

I have reviewed the code, committed some minor changes to use namespace directly.  edited I/O reference and PR description.  

I think it is good to go, but I will leave it to someone else to merge when it is a good time for that.

Myoldmopar commented:
Is this branch way out of date with develop?  Or did something go awry with that last merge/commit?

lgu1234 commented:
@Myoldmopar This branch is based on energy-plus repository. I dont' know how to handle it. I can merge it with energy-plus repository.

Myoldmopar committed:
Merge remote-tracking branch 'upstream/develop' into EliminateIteration-Release

lgu1234 commented:
@Myoldmopar Shall I need to merge it?

Myoldmopar commented:
Do not merge it.  The last results had lots of diffs and CI will need to run on it first.  If it's all clean we'll get it merged tomorrow.

Myoldmopar commented:
Yeah that's maybe a little cleaner :laughing:
